core/src/main/scala/chisel3/ElaboratedCircuit.scala

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3
+
+import firrtl.annotations.Annotation
+import firrtl.ir.{CircuitWithAnnos, Serializer}
+import firrtl.options.{CustomFileEmission, Unserializable}
+import chisel3.experimental.{BaseModule, SourceInfo, UnlocatableSourceInfo}
+import chisel3.experimental.hierarchy.Definition
+import chisel3.internal.firrtl.ir.Circuit
+import chisel3.internal.firrtl.Converter
+
+/** The result of running Chisel elaboration
+  *
+  * Provides limited APIs for inspection of the resulting circuit.
+  */
+// This is an important interface so let's keep it separate from the implementation.
+sealed trait ElaboratedCircuit {
+
+  /** The name of the circuit, also the name of the top public module */
+  def name: String
+
+  /** The circuit as a string of FIRRTL IR
+    *
+    * This will include annotations automatically.
+    * For large circuits (> 2 GiB of text) use [[lazilySerialize]].
+    */
+  def serialize: String
+
+  /** The circuit as a string of FIRRTL IR
+    *
+    * For large circuits (> 2 GiB of text) use [[lazilySerialize]]
+    *
+    * @param annotations alternative annotations to include in the FIRRTL IR.
+    */
+  def serialize(annotations: Iterable[Annotation]): String
+
+  /** The circuit as a String of FIRRTL IR
+    *
+    * This will include annotations automatically.
+    * Serialized lazily to reduce peak memory use and support cicuits larger than 2 GiB.
+    */
+  def lazilySerialize: Iterable[String]
+
+  /** The circuit as a String of FIRRTL IR
+    *
+    * Serialized lazily to reduce peak memory use and support cicuits larger than 2 GiB.
+    *
+    * @param annotations alternative annotations to include in the FIRRTL IR.
+    */
+  def lazilySerialize(annotations: Iterable[Annotation]): Iterable[String]
+
+  /** The annotations created during elaboration of this circuit
+    *
+    * This does not include annotations passed to elaboration.
+    */
+  def annotations: Iterable[Annotation]
+
+  /** Definitions of all modules in this circuit */
+  def definitions: Iterable[Definition[BaseModule]]
+
+  /** The underlying circuit, for private use only */
+  private[chisel3] def _circuit: Circuit
```

Comment:
Nit: `circuit` is probably fine given that this is private.

Reply:
After reading further, I see that this may be a useful convention throughout. Do whatever makes sense to you.

Reply:
I'm going to stick with this for now but since it's package private we can change it arbitrarily.